accepted/2021/System.Diagnostics/Metrics-Design.md

```diff
@@ -0,0 +1,484 @@
+    public abstract class ObservableInstrument<T> : Instrument where T : unmanaged
+    {
+        /// <summary>
+        /// Protected constructor to create the instrument with the common properties.
+        /// </summary>
+        protected ObservableInstrument(
+                    Meter meter,
+                    string name,
+                    string? description,
+                    string? unit) : base(meter, name, description, unit) { throw null; }
+
+        /// <summary>
+        /// Observe() fetches the current measurements being tracked by this instrument.
+        /// </summary>
+        protected abstract IEnumerable<Measurement<T>> Observe();
+
+        public override bool IsObservable => throw null;
+    }
+}
+```
+
+### Measurement class
+
+```csharp
+
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// A measurement stores one observed value and its associated tags. This type is used by Observable instruments' Observe() method when reporting current measurements.
+    /// with the associated tags.
+    /// </summary>
+    public struct Measurement<T> where T : unmanaged
+    {
+        /// <summary>
+        /// Construct the Measurement using the value and the list of tags.
+        /// We'll always copy the input list as this is not perf hot path.
+        /// </summary>
+        public Measurement(T value, IEnumerable<KeyValuePair<string, object?>> tags) { throw null; }
+        public Measurement(T value, params KeyValuePair<string, object?>[] tags) { throw null; }
+        public Measurement(T value, ReadOnlySpan<KeyValuePair<string, object?>> tags) { throw null; }
+
+        public ReadOnlySpan<KeyValuePair<string, object?>> Tags { get { throw null;  } }
+        public T Value { get; }
+    }
+}
+```
+
+### Non-observable Instruments
+
+```csharp
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// The counter is a non-observable Instrument that supports non-negative increments.
+    /// e.g. Number of completed requests.
+    /// </summary>
+    public sealed class Counter<T> : Instrument<T> where T : unmanaged
+    {
+        public void Add(T measurement) { throw null; }
+        public void Add(T measurement,
+                            KeyValuePair<string, object?> tag1) { throw null; }
+        public void Add(T measurement,
+                            KeyValuePair<string, object?> tag1,
+                            KeyValuePair<string, object?> tag2) { throw null; }
+        public void Add(T measurement,
+                            KeyValuePair<string, object?> tag1,
+                            KeyValuePair<string, object?> tag2,
+                            KeyValuePair<string, object?> tag3) { throw null; }
+        public void Add(T measurement,
+                            ReadOnlySpan<<string, object?>> tags) { throw null; }
+        public void Add(T measurement,
+                            params KeyValuePair<string, object?>[] tags) { throw null; }
```

Comment:
Assuming our end result is ReadOnlySpan<KeyValuePair<>>, I'm wondering if the 1,2,3 KVP prototypes is any faster than a proto that takes param KVP. If it's not faster, then, I don't know we need these 1,2,3 KVP prototypes.

Reply:
Yes. But if our goal is to create ReadOnlySpan<>, we will still need to create an array! So, practically, I'm wondering if there's any difference at this point!

Reply:
> Yes. But if our goal is to create ReadOnlySpan<>, we will still need to create an array! So, practically, I'm wondering if there's any difference at this point!

No, we don't need to create array for that. we'll just store the values on the stack and create the ReadOnlySpan from that.